src/vs/workbench/services/files/node/fileService.ts

```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
 		return this.doResolveContent(resource, options, (stat, enc) => this.resolveFileContent(stat, enc));
 	}
 
+	private static chunkSize = 64 * 1024;
+	private static chunkBuffer = Buffer.allocUnsafe(FileService.chunkSize);
+
+	public resolveStringStream(resource: uri, options?: IResolveContentOptions, token: CancellationToken = CancellationToken.None): IStringStream {
+		const ret = new EventEmitter();
+
+		fs.open(resource.path, 'r', (err, fd) => {
+
+			if (err) {
+				ret.emit('error', err);
+				return;
+			}
+
+			let decoder: NodeJS.ReadWriteStream;
+			let totalBytesRead = 0;
+
+			const finish = (err?: any) => {
+				if (err) {
+					ret.emit('error', err);
+				}
+				fs.close(fd, err => {
+					if (err) {
+						ret.emit('error', err);
+					}
+					if (decoder) {
+						decoder.removeAllListeners();
+					}
+					ret.emit('end');
+				});
+			};
+
+			const handleChunk = (bytesRead) => {
+				if (token.isCancellationRequested) {
+					// cancellation
+					finish(new Error('cancelled'));
+
+				} else if (bytesRead < FileService.chunkSize) {
+					// done, write rest, end
+					decoder.write(FileService.chunkBuffer.slice(0, bytesRead), finish);
+
+				} else {
+					// read, write, repeat
+					decoder.write(FileService.chunkBuffer, readChunk);
+				}
+			};
+
+			const readChunk = () => {
+				fs.read(fd, FileService.chunkBuffer, 0, FileService.chunkSize, null, (err, bytesRead) => {
+					totalBytesRead += bytesRead;
+
+					if (totalBytesRead > MAX_FILE_SIZE) {
+						// reading too much... this looks stupid because we have already read much
+						// but it safes an upfront stat call for every 'normal' file we open.
+						finish(new FileOperationError(
+							nls.localize('fileTooLargeError', "File too large to open"),
+							FileOperationResult.FILE_TOO_LARGE
+						));
+					} else if (err) {
+						// some error happened
+						finish(err);
+
+					} else if (decoder) {
+						//
+						handleChunk(bytesRead);
+
+					} else {
+						// when receiving the first chunk of data we need to create the
+						// decoding stream which is then used to drive the string stream.
+						Promise.resolve(detectMimeAndEncodingFromBuffer(
+							{ buffer: FileService.chunkBuffer, bytesRead },
+							options && options.autoGuessEncoding)
+						).then(value => {
+
+							if (options && options.acceptTextOnly && value.mimes.indexOf(baseMime.MIME_BINARY) >= 0) {
+								// Return error early if client only accepts text and this is not text
+								finish(new FileOperationError(
+									nls.localize('fileBinaryError', "File seems to be binary and cannot be opened as text"),
+									FileOperationResult.FILE_IS_BINARY
+								));
+
+							} else {
```

Comment:
A lots of instructions within `else`...
Make it more readable and simple.
Put some logic to methods.